### PR TITLE
Move parallelcompat and chunkmanagers to NamedArray

### PR DESCRIPTION
**Pull request by TomNicholas**
@dcherian I got to this point before realizing that simply moving `parallelcompat.py` over isn't [what it says in the design doc](https://github.com/pydata/xarray/blob/main/design_notes/named_array_design_doc.md#appendix-implementation-details), which instead talks about

> - Could this functionality be left in Xarray proper for now? Alternative array types like JAX also have some notion of "chunks" for parallel arrays, but the details differ in a number of ways from the Dask/Cubed.
> - Perhaps variable.chunk/load methods should become functions defined in xarray that convert Variable objects. This is easy so long as xarray can reach in and replace `.data`

I personally think that simply moving parallelcompat makes sense so long as you expect people to use chunked `NamedArray` objects. I see the chunked arrays as special cases of duck arrays, and my understanding is that `NamedArray` is supposed to have full support for duckarrays.

cc @andersy005 

- [x] As requested in #8238
- [ ] ~~Tests added~~
- [ ] ~~User visible changes (including notable bug fixes) are documented in `whats-new.rst`~~
- [ ] ~~New functions/methods are listed in `api.rst`~~

### CODE REVIEW

**Comment by andersy005** [comment]
thank you for doing this work, @TomNicholas! moving these modules unblock me. 

> I personally think that simply moving parallelcompat makes sense so long as you expect people to use chunked NamedArray objects. I see the chunked arrays as special cases of duck arrays, and my understanding is that NamedArray is supposed to have full support for duckarrays.

👍🏽 i concur. and yes, `NamedArray` will have full support for duckarrays.

**Comment by TomNicholas** [comment]
Moving `utils.either_dict_or_kwargs` to `namedarray.utils` required touching a lot of files to fix!

Also moving the entrypoint seems to have broken the discovery of the `DaskManager` at least in my local environment, let's see what happens in the CI...

File: xarray/namedarray/core.py
Comment (TomNicholas) [inline xarray/namedarray/core.py:635]: Including `.load` on `NamedArray` means we would need to move `as_compatible_data` to `namedarray`, but that depends on a nuch of stuff including `duck_array_ops`. Maybe I should leave `.load` out for now @andersy005 ?

Comment (TomNicholas) [reply xarray/namedarray/core.py:635]: Okay I got stuck here. I tried moving `.load`, `as_compatible_data` etc. to `xarray.namedarray.core`, but that requires `NON_NUMPY_SUPPORTED_ARRAY_TYPES` which requires `pd.Index` (as well as other pandas stuff). Presumably you don't want any pandas imports in the namedarray module, so what should I do here? 

I could import `as_compatible_data` from xarray somewhere, but it can't live in `xarray.core.variable` else I will probably have a circular import. This also wouldn't really solve the pandas dependency. Another option might be to use a less stringent coercion that `as_compatible_data` inside `NamedArray.load`.

File: xarray/namedarray/core.py
Comment (andersy005) [inline xarray/namedarray/core.py:983]:
```suggestion
```

i believe we already have this in place, 

https://github.com/pydata/xarray/blob/eb749448f9513517230090b2cb0e3521ccb4f20f/xarray/namedarray/core.py#L756

File: xarray/namedarray/pycompat.py
Comment (andersy005) [inline xarray/namedarray/pycompat.py:11]: there's a version of `module_availabe` in `namedarray.utils`

```suggestion
from xarray.core.utils import is_scalar
from xarray.namedarray.utils import is_duck_array, module_available
```

File: xarray/namedarray/utils.py
Comment (andersy005) [inline xarray/namedarray/utils.py:61]: i believe all of these are currently available from `namedarray/_typing.py`



```suggestion

```

File: xarray/namedarray/utils.py
Comment (andersy005) [inline xarray/namedarray/utils.py:177]: these are also available in `namedarray/_array_api.py`

https://github.com/pydata/xarray/blob/eb749448f9513517230090b2cb0e3521ccb4f20f/xarray/namedarray/_array_api.py#L38


```suggestion
```

**Comment by TomNicholas** [comment]
Thanks for the review @andersy005. I feel like I'm trying to hit a moving target with this PR - I'm pretty sure all your comments would not have applied last week. I'll try to make it compatible now though.

[This pandas dependency issue](https://github.com/pydata/xarray/pull/8319#discussion_r1364360043) however is an actual problem - I don't know how to proceed there.

**Comment by andersy005** [comment]
thank you, @TomNicholas! i have plenty of time today, and i'm happy to help bring this PR to completion.

**Comment by TomNicholas** [comment]
> i have plenty of time today, and i'm happy to help bring this PR to completion.

Great!

Ping me if you need help here @andersy005

File: xarray/namedarray/_typing.py
Comment (Illviljan) [inline xarray/namedarray/_typing.py:58]: I think we can use either or here. I'm not convinced that using TypeVar in _Chunks is the best idea either.

File: xarray/namedarray/core.py
Comment (Illviljan) [inline xarray/namedarray/core.py:729]: A lot of typing here. Do we like them all?

Comment (andersy005) [reply xarray/namedarray/core.py:729]: Ccing @TomNicholas

File: xarray/namedarray/utils.py
Comment (andersy005) [reply xarray/namedarray/utils.py]: Done!

File: xarray/namedarray/core.py
Comment (andersy005) [reply xarray/namedarray/core.py:736]: Ccing @TomNicholas

**Comment by TomNicholas** [comment]
@Illviljan and @andersy005 I think we should decide whether to try and fix the (valid) issues with input arguments to `.chunks` etc. in this PR or a later PR.

For context, these method signatures are somewhat messy because they were dask-specific, they built up various dask-specific kwargs over time, then I generalized them to handle anything via `from_array_kwargs`, but currently the main xarray API has both for backwards-compatibility purposes.

File: xarray/namedarray/core.py
Comment (TomNicholas) [reply xarray/namedarray/core.py:736]: We can simply this a alot. It should be possible to remove `name`, `lock`, and `inline_array` in favour of them being passed via `from_array_kwargs`, because they are just dask-specific kwargs to `ChunkManager.from_array`. IIRC they already can be passed this way, they are just also explicitly present in the signature for backwards compatibility.

File: xarray/namedarray/core.py
Comment (TomNicholas) [reply xarray/namedarray/core.py:729]: I think given https://github.com/pydata/xarray/pull/8341 we can remove the two `tuple` types. However as the deprecation cycle for passing a tuple of chunks to eg. `DataArray.chunk` has only just begun, I think we will need to add a piece of code to change tuples to mappings of chunks in the same place as the deprecation warning is raised in #8341. (cc @max-sixty ) We still need the int, Literal, and Mapping.

File: xarray/namedarray/core.py
Comment (TomNicholas) [reply xarray/namedarray/core.py:736]: Chunks should be positional, but I don't think any of the others need be.

File: xarray/namedarray/core.py
Comment (Illviljan) [reply xarray/namedarray/core.py:729]: The deprecation cycle is for DataArray/Variable though. So I think we should try to keep NamedArray as close to the ideal as possible, and solve the backwards compat issues in DataArray/Variable.

I think we should be careful inheriting api that we aren't satisfied with, it's easy to forget once it's merged.

Comment (TomNicholas) [reply xarray/namedarray/core.py:729]: > I think we should be careful inheriting api that we aren't satisfied with, it's easy to forget once it's merged.

That's fair enough.

> The deprecation cycle is for DataArray/Variable though.

I think we agree - I'm just trying to make it clear that having these arguments in DataArray/Variable but not in NamedArray requires some compatibility code that hasn't been written yet, at least until the deprecation cycle is finished.

File: xarray/namedarray/core.py
Comment (andersy005) [reply xarray/namedarray/core.py]: considering how large this PR has gotten, should we consider addressing these additional issues in separate PRs?

Comment (andersy005) [reply xarray/namedarray/core.py]: never mind :) Tom pointed it out below: https://github.com/pydata/xarray/pull/8319#issuecomment-1781225864

**Comment by andersy005** [comment]
@dcherian / @TomNicholas / @Illviljan, this is ready for another round of review..

i might have overlooked/ignored quite a few typing errors, but I'm more than willing to correct them with a bit of assistance ;)

File: xarray/core/variable.py
Comment (dcherian) [inline xarray/core/variable.py:941]: Can we use `self._data = to_duck_array(self._data)` here instead?

File: xarray/namedarray/core.py
Comment (dcherian) [inline xarray/namedarray/core.py:639]: These should be raising errors now since we require that error type be specified for `type: ignore`

File: xarray/namedarray/core.py
Comment (dcherian) [inline xarray/namedarray/core.py:873]: I think we have a `to_numpy` utility function with this logic.

File: xarray/core/dataset.py
Comment (TomNicholas) [inline xarray/core/dataset.py:118]: It doesn't look like we need to import this here? (Also if dask is not installed can we even import it?)

File: xarray/namedarray/core.py
Comment (TomNicholas) [reply xarray/namedarray/core.py:742]: I don't think this is actually resolved? Unless it's being deferred to a later PR?

File: xarray/namedarray/core.py
Comment (TomNicholas) [inline xarray/namedarray/core.py:745]: Talking about other API that isn't satisfying - `name`, `lock`, `inline_array` are all dask-specific, so should ideally be folded into `from_array_kwargs`

File: xarray/namedarray/daskmanager.py
Comment (TomNicholas) [inline xarray/namedarray/daskmanager.py:61]: Is there a reason why we're setting all these types to `Any` here?

File: xarray/namedarray/parallelcompat.py
Comment (TomNicholas) [inline xarray/namedarray/parallelcompat.py:36]: I like this general idea - could we make it something like

```python
class ChunkedArrayMixin(Protocol):
    def rechunk(self, chunks: Any, **kwargs: Any) -> Any: ...

    @property
    def chunks(self) -> _NormalizedChunks: ...

    @property
    def compute(self) -> np.ndarray: ...

File: xarray/namedarray/utils.py
Comment (TomNicholas) [inline xarray/namedarray/utils.py:227]: This whole function might not need to live in `NamedArray`, just in xarray main for backwards compatibility. See my other comment about removing dask-specific kwargs from the NamedArray API.

File: xarray/namedarray/daskmanager.py
Comment (andersy005) [reply xarray/namedarray/daskmanager.py:61]: originally, i was trying to appease `mypy` but i'm going to try to reintroduce them in a separate commit

Comment (andersy005) [reply xarray/namedarray/daskmanager.py]: this has been addressed

**Comment by andersy005** [comment]
this is ready for another round of review :)

**Review by TomNicholas — COMMENTED** [review]
I can't approve my own PR, but I approve! :+1:

**Review by andersy005 — APPROVED** [review]

**Pull request merged**
